### PR TITLE
DNS: Don't try to apply empty changesets

### PR DESCRIPTION
**Pull request by justinsb**
For GCE this was resulting in confusing error messages.

### CODE REVIEW

File: dnsprovider/pkg/dnsprovider/providers/google/clouddns/rrchangeset.go [hunk 1]
```diff
@@ -115,7 +119,7 @@ func (c *ResourceRecordChangeset) Apply() error {
 }
 
 func (c *ResourceRecordChangeset) IsEmpty() bool {
-	return len(c.additions) == 0 && len(c.removals) == 0
+	return len(c.additions) == 0 && len(c.removals) == 0 && len(c.upserts) == 0
```
Comment (johngmyers) [inline dnsprovider/pkg/dnsprovider/providers/google/clouddns/rrchangeset.go]: Shouldn't there be a unit test for this, covering all of the implementations of the interface method? The only prior caller to this method I could find was a unit test for the case of it being empty. There should be tests for one addition, one removal, and one upsert.

Comment (justinsb) [reply dnsprovider/pkg/dnsprovider/providers/google/clouddns/rrchangeset.go:122]: Yes, there should be.  We inherited a lot of this code from k/k, and the design is pretty different from the rest of the code (even the rest of the code in k/k).  It's hard to resist just rewriting the whole thing...

**Comment by justinsb** [comment]
It should be the second contract, I believe.  I'll add a comment and fix up the others!

**Comment by justinsb** [comment]
Documented the contract, changed the signature to force a code change to every call & definition, added a generic test of the contract behaviour that was problematic.

**Comment by johngmyers** [comment]
Looks good, aside from the gofmt failure.

**Pull request merged**